cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,336 @@
+  /**
+   * Update goal state after one round of rebalance.
+   * @param clusterModel The state of the cluster.
+   * @param excludedTopics The topics that should be excluded from the optimization proposal.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, Set<String> excludedTopics) {
+    // Log broker Ids over balancing limit.
+    // While proposals exclude the excludedTopics, the balance still considers utilization of the excludedTopic replicas.
+    if (!_brokerIdsAboveBalanceUpperLimit.isEmpty()) {
+      LOG.warn("Leader replica count on broker ids:{} {} above the balance limit of {} after rebalancing.",
+               _brokerIdsAboveBalanceUpperLimit, (_brokerIdsAboveBalanceUpperLimit.size() > 1) ? "are" : "is",
+               _balanceUpperLimit);
+      _succeeded = false;
+    }
+    if (!_brokerIdsUnderBalanceLowerLimit.isEmpty()) {
+      LOG.warn("Leader replica count on broker ids:{} {} under the balance limit of {} after rebalancing.",
+               _brokerIdsUnderBalanceLowerLimit, (_brokerIdsUnderBalanceLowerLimit.size() > 1) ? "are" : "is",
+               _balanceLowerLimit);
+      _succeeded = false;
+    }
+    finish();
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   *
+   * @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization -- e.g. excluded topics.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) {
+    LOG.debug("Rebalancing broker {} [limits] lower: {} upper: {}.", broker.id(), _balanceLowerLimit, _balanceUpperLimit);
+    int numLeaderReplicas = broker.leaderReplicas().size();
+    // Update broker ids over the balance limit for logging purposes.
+    if (numLeaderReplicas > _balanceUpperLimit && rebalanceByMovingLeadershipOut(broker, clusterModel, optimizedGoals, optimizationOptions)) {
+      _brokerIdsAboveBalanceUpperLimit.add(broker.id());
+      LOG.debug("Failed to sufficiently decrease leader replica count in broker {} with leadership movements. Leader replicas: {}.",
+                broker.id(), broker.leaderReplicas().size());
+    } else if (numLeaderReplicas < _balanceLowerLimit && rebalanceByMovingLeadershipIn(broker, clusterModel, optimizedGoals, optimizationOptions)) {
+      _brokerIdsUnderBalanceLowerLimit.add(broker.id());
+      LOG.debug("Failed to sufficiently increase leader replica count in broker {} with leadership movements. Leader replicas: {}.",
+                broker.id(), broker.leaderReplicas().size());
+    }
+  }
+
+  private boolean rebalanceByMovingLeadershipOut(Broker broker,
```

Comment:
Why do we only allow leadership moves? If leadership move doesn't work, can't we allow (1) replica moves and (2) replica swaps?

Reply:
I added leader replica move operation